File: app/src/main/java/com/norvera/confession/ui/guide/GuideListFragmentAdapter.java
package com.norvera.confession.ui.guide;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.norvera.confession.data.models.GuideEntry;
import com.norvera.confession.databinding.FragmentGuideEntryBinding;

import androidx.annotation.NonNull;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

public class GuideListFragmentAdapter extends ListAdapter<GuideEntry, GuideListFragmentAdapter.ViewHolder> {
    GuideListFragmentAdapter() {
        super(DIFF_CALLBACK);
    }

    private static final DiffUtil.ItemCallback<GuideEntry> DIFF_CALLBACK = new DiffUtil.ItemCallback<GuideEntry>() {
        @Override
        public boolean areItemsTheSame(@NonNull GuideEntry guideEntry, @NonNull GuideEntry t1) {
            return guideEntry.id == t1.id;
        }

        @Override
        public boolean areContentsTheSame(@NonNull GuideEntry guideEntry, @NonNull GuideEntry t1) {
            return guideEntry.equals(t1);
        }
    };

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int i) {
        FragmentGuideEntryBinding binding = FragmentGuideEntryBinding.inflate(
                LayoutInflater.from(parent.getContext()), parent, false);

        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        GuideEntry guideEntry = getItem(position);
        holder.onBind(createOnClickListener(guideEntry.id, guideEntry), guideEntry);

    }

    private View.OnClickListener createOnClickListener(long id, GuideEntry guideEntry) {
        return view -> {
            GuideListFragmentDirections.ActionGuideFragmentListToGuideFragmentDetail toGuideFragmentDetail =
                    GuideListFragmentDirections.actionGuideFragmentListToGuideFragmentDetail(guideEntry);
            toGuideFragmentDetail.setTitle(guideEntry.guideTitle);
            toGuideFragmentDetail.setGuideEntry(guideEntry);
            Navigation.findNavController(view).navigate(toGuideFragmentDetail);
        };
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        private FragmentGuideEntryBinding binding;

        ViewHolder(@NonNull FragmentGuideEntryBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

            void onBind(View.OnClickListener clickListener, GuideEntry item) {
                itemView.setTag(item);
                binding.setClickListener(clickListener);
                binding.setGuideEntry(item);
                binding.executePendingBindings();



        }
    }
}

File: app/src/main/java/com/norvera/confession/ui/guide/GuideListFragmentAdapter.kt
package com.norvera.confession.ui.guide

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.navigation.Navigation.findNavController
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.norvera.confession.data.models.GuideEntry
import com.norvera.confession.databinding.FragmentGuideEntryBinding

class GuideListFragmentAdapter internal constructor() :
    ListAdapter<GuideEntry, GuideListFragmentAdapter.ViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, i: Int): ViewHolder {
        val binding = FragmentGuideEntryBinding.inflate(
            LayoutInflater.from(parent.context), parent, false
        )

        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val guideEntry = getItem(position)
        holder.onBind(createOnClickListener(guideEntry.id, guideEntry), guideEntry)

    }

    private fun createOnClickListener(id: Long, guideEntry: GuideEntry): View.OnClickListener {
        return View.OnClickListener{
            val toGuideFragmentDetail =
                GuideListFragmentDirections.actionGuideFragmentListToGuideFragmentDetail(guideEntry.guideTitle, guideEntry)
            findNavController(it).navigate(toGuideFragmentDetail)
        }
    }

    inner class ViewHolder(private val binding: FragmentGuideEntryBinding) :
        RecyclerView.ViewHolder(binding.root) {

        fun onBind(clickListener: View.OnClickListener, item: GuideEntry) {
            itemView.tag = item
            binding.clickListener = clickListener
            binding.guideEntry = item
            binding.executePendingBindings()


        }
    }

    companion object {

        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<GuideEntry>() {
            override fun areItemsTheSame(guideEntry: GuideEntry, t1: GuideEntry): Boolean {
                return guideEntry.id == t1.id
            }

            override fun areContentsTheSame(guideEntry: GuideEntry, t1: GuideEntry): Boolean {
                return guideEntry == t1
            }
        }
    }
}
